Disable tethering while tooltips are hidden

joukevandermaas opened the pull request:
This was causing performance issues on pages with many tooltips, because each repositioning triggered a forced reflow (which is expensive in most browsers, but especially in IE).

joukevandermaas commented:
I can't figure out why the build is failing, but I'm pretty sure it's not (entirely) related to this PR. Can someone maybe point me in the right direction?

a15n commented on addon/components/tooltip-on-element.js:
```diff
@@ -336,6 +338,10 @@ export default EmberTetherComponent.extend({
 
     this.set('offset', offset);
+
+    if (event !== 'none' || this.get('tooltipIsVisible') !== true) {
```
nit: `!this.get('tooltipIsVisible')`

a15n commented on addon/components/tooltip-on-element.js:
```diff
@@ -430,6 +440,16 @@ export default EmberTetherComponent.extend({
   },
 
+  stopTether() {
+    run.schedule('afterRender', this, function() {
```
nit: if you use arrow functions you won't need to pass in `this`

```
run.schedule('afterRender', () => {
  ...
});
```

a15n commented on addon/components/tooltip-on-element.js:
```diff
@@ -430,6 +440,16 @@ export default EmberTetherComponent.extend({
+
+  startTether() {
+    this.get('_tether').enable();
+  },
```
could you add tests to check that the tether is enabled when we expect it to be?

```
isTetherEnabled: computed.alias('_tether.enabled'),
```

andrew-zenefits replied:
We copied your fix here at Zenefits and did this to get test coverage

https://github.com/zenefits/ember-tooltips/pull/21/files#diff-eafc17148ee89489b3d8098122b506e6R409

sir-dunxalot commented:
As Andrew said, this is a great fix. Good catch and thank you for putting in the time to issue the PR. I will wait until #94 has landed to merge this but the code looks good so I don't have any comment there. This will likely need a quick rebase to resolve the merge conflicts once #94 is merged onto master.

joukevandermaas commented:
OK, I'll address @a15n's comments and rebase after #94 is merged, then. Thanks!

a15n commented:
@joukevandermaas I think this may be helpful. It has full test coverage. I don't know if what I did with an `is-tether-enabled` attribute is ideal. A 'js-' prefixed class may be better. 

https://github.com/zenefits/ember-tooltips/compare/v2.0.0-zenefits.1...v2.0.0-zenefits.2

joukevandermaas commented:
I see some refactoring happened in the meantime, while adding popovers. I hope to get this PR fixed up today or tomorrow if I can.

sir-dunxalot commented:
Sounds great, Jouke! Let myself or @a15n know if you have any questions about the refactoring from popovers.

joukevandermaas commented:
I rebased, adapted the code and added some asserts to the visibility tests. I squashed too, so the 'old' code is no longer visible here.

I don't know why the beta build fails, but it doesn't look like it's anything to do with this PR.

a15n commented on tests/integration/components/tooltip-is-visible-test.js:
```diff
@@ -1,6 +1,6 @@
 import Ember from 'ember';
 import { moduleForComponent, test } from 'ember-qunit';
-import { assertHide, assertShow } from '../../helpers/sync/assert-visibility';
+import { assertHide, assertShow, assertTetherEnabled, assertTetherDisabled } from '../../helpers/sync/assert-visibility';
```
I would recommend against creating a new assertion just for the tether. This means that people will always have to opt in to the usage and will likely skip that test. Instead I would add the check as part of assertShow and assertHide. When I did this I found some holes in my enable/disable coverage

https://github.com/zenefits/ember-tooltips/pull/21/files#diff-c2e9fd979e3b24030663b2c4c84cd5beR3

a15n replied:
I think this should be blocking

joukevandermaas replied:
Yeah, good point. I'll try split it (or make a new PR if you want to merge already).

a15n commented on addon/components/tooltip-and-popover.js:
```diff
@@ -419,4 +435,20 @@ export default EmberTetherComponent.extend({
 
+  startTether() {
+    // can't depend on `_tether.enabled` because it's not an
+    // Ember property (so won't trigger cp update when changed)
```
good note

a15n commented:
the two test failures are expected and not related to these changes. @sir-dunxalot maybe we should make the `ember-beta` test failure allowable?

lvl99 commented:
Pretty keen to see this integrated with the main build. Currently have some extreme lag after adding tooltips to my Ember app.

sir-dunxalot commented:
Yes, I think making `ember-beta` failure allowable is acceptable given the conversations about the impact of glimmer (see #97). As such, this PR should be good to go once @a15n 's code comments have been addressed.

joukevandermaas commented:
@a15n I moved the tethering assert into the existing helper. There was a slight complication, because for delayed tooltips the tethering should be enabled but the tooltip should remain hidden for a while. I added some flag to the helper now, but another potential solution is to split this specific scenario into a new helper. Let me know what you think.

a15n commented:
@joukevandermaas thanks for your work on this.

When we did this for our local repo we didn't have to provide any additional parameters for the tests. I think that if you add the `startTether()` call inside the if/else within the show function you won't need the additional test parameter or special handling. 

https://github.com/zenefits/ember-tooltips/pull/21/files#diff-eafc17148ee89489b3d8098122b506e6R382

joukevandermaas commented:
Good point, I've updated the code to do that and the special assert is no longer needed. Thanks!

a15n commented:
Great work @joukevandermaas! This code LGTM.

@sir-dunxalot ^

sir-dunxalot commented:
+1. I will merge this, move `ember-beta` to allowed failures, and will release today with the passing test suite.

Great work all.

sir-dunxalot commented:
Releases as `2.3.1`.
